tools/crudd-speed-battery.sh

```diff
@@ -0,0 +1,183 @@
+#!/bin/bash
+
+set -euo pipefail
+
+REGION_SIZE_MIBS=4096
+REGION_SIZE_BYTES=$(( REGION_SIZE_MIBS * 1024 * 1024 ))
+BLOCK_SIZE=512
+REGION_SIZE_BLOCKS=$(( REGION_SIZE_BYTES / BLOCK_SIZE ))
+# 128 MiB
+EXTENT_SIZE_BLOCKS=$(( 128 * 1024 * 1024 / BLOCK_SIZE ))
+EXTENT_COUNT=$(( REGION_SIZE_BLOCKS / EXTENT_SIZE_BLOCKS ))
+
+MAX_ITERATION_DURATION=$((2 * 60))
+
+BENCHMARK_RESULTS_TMPFILE=/tmp/crudd-benchmark-num-bytes-processed.txt
+
+DOWNSTAIRS_1=127.0.0.1:8810
+DOWNSTAIRS_2=127.0.0.1:8820
+DOWNSTAIRS_3=127.0.0.1:8830
+
+print_err() {
+  printf '%s\n' "$1" 1>&2
+}
+
+# $1 = mode (read or write)
+# $2 = request size in blocks
+# $3 = pipeline length
+# runs for a maximum time specified by the use of timeout in the function
+# prints the average throughput in bytes/second to stdout.
+crudd_benchmark() {
+  local mode="$1"
+  local req_size="$2"
+  local pipeline_len="$3"
+
+  local start_time="$(date '+%s')"
+
+  # SIGUSR1 is handled specially by crudd in bechmark mode
+  # In benchmark mode, crudd prints how many bytes it wrote to
+  # the file passed as an argument to --benchmarking-mode. the timeout
+  # ensures crudd never runs _longer_ than the provided duration, but it could
+  # run shorter if we're fast and the region is small. So we take bytes written,
+  # divided by the time, to get our actual speed. This is all a little bit jank,
+  # admittedly.
+  timeout --signal=USR1 "$MAX_ITERATION_DURATION" \
+    "$BINDIR/crudd" -t $DOWNSTAIRS_1 -t $DOWNSTAIRS_2 -t $DOWNSTAIRS_3 \
+      -n $REGION_SIZE_BYTES -i "$req_size" -p "$pipeline_len" \
+      --benchmarking-mode "$BENCHMARK_RESULTS_TMPFILE" \
+      "$mode" \
+      1>&2
+
+  local end_time="$(date +'%s')"
+  local bytes_processed="$(cat "$BENCHMARK_RESULTS_TMPFILE")"
+
+  # We lose some precision here, but with speeds in the KiB/s or MiB/s, the
+  # rounding loss is not much concern
+  printf '%s' $((bytes_processed / (end_time - start_time) ))
+}
+
+# create and start downstairs with dsc
+create_and_start_downstairs() {
+  if [[ -n ${DSC_PID-} ]]; then
```

Comment:
Should we check that `$DSC_PID` is running like on line 83?  This just tells us that
the variable is set which may or may not indicate a running dsc.

Reply:
It couldn't hurt. The point of this check was to make sure I didn't call `create_and_start_downstairs` unless I had previously called `stop_downstairs` (which clears the variable). But we could add some more verification of the environment while we're at it.